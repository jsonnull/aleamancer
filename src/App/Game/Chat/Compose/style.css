@import 'base';
@import 'input';

.darkTheme {}
.unpinned {}

.form {
  padding: 2.4rem 1.2rem;
  padding-top: 10px;
  border-radius: 5px;
}
  .formInput {
    @extend input;
    border: 0;
    height: 50px;
    display: block;
    padding: 1.2rem 1rem;
    line-height: 2.4rem;
    width: 100%;
    resize: none;
    color: $fgPrimary;
    border: 1px solid $borderColorLight;
    border-radius: 5px;

    :global(.dark) & {
      color: $fgPrimaryDark;
      background-color: $bgPrimaryDark;
      border: 1px solid $borderColorDark;
    }

    :global(.unpinned) & {
      background-color: $bgPrimary;
      box-shadow: 0 3px 5px 0 rgba(0, 0, 0, 0.2);
      border: 1px solid white;

      :global(.dark) & {
        background-color: $bgPrimaryDark;
      }
    }
  }
