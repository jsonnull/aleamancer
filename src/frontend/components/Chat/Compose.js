// @flow
import React from 'react'
import styled from 'styled-components'
import { fontSize, fonts } from 'frontend/styles/common'

const Form = styled.form``

const Input = styled.textarea.attrs({
  type: 'text'
})`
  border: 0;
  height: 50px;
  line-height: 3rem;
  padding: 1rem;
  display: block;
  width: 100%;
  resize: none;
  color: ${props => props.theme.color};
  background-color: ${props => props.theme.background};
  border-radius: 5px;
  font-size: ${fontSize.normal};
  font-family: ${fonts.body};
`

type Props = {
  onSend: Function,
  isPinned: boolean
}

type State = {
  value: string,
  height: number
}

const defaultHeight = 50

export default class Compose extends React.Component<Props, State> {
  focused: boolean
  messageQueue: Array<string>
  autogrow: ?HTMLElement

  constructor(props: Props) {
    super(props)

    this.state = {
      value: '',
      height: 0
    }
    this.focused = false
    this.messageQueue = []
  }

  handleChange(event: Object) {
    const value = event.target.value

    const setHeightImmediate = value.length > this.state.value.length

    this.setState({
      value
    })

    if (setHeightImmediate) {
      this.calculateHeight()
    } else {
      this.setState({ height: 0 })
      window.requestAnimationFrame(() => this.calculateHeight())
    }
  }

  calculateHeight() {
    if (!this.autogrow) {
      return
    }

    const autogrow = this.autogrow
    const scrollHeight = autogrow.scrollHeight

    let height = 0
    if (scrollHeight !== defaultHeight) {
      height = scrollHeight
    }

    this.setState({
      height
    })
  }

  handleKeyUp(event: Object) {
    if (event.key == 'Enter') {
      this.handleSubmit()
    }
  }

  handleSubmit() {
    this.messageQueue.push(this.state.value)
    this.props.onSend(this.state.value)
    this.setState({
      value: '',
      height: 0
    })
  }

  render() {
    const { isPinned } = this.props

    const style = {}
    if (this.state.height !== 0) {
      Object.assign(style, { height: this.state.height + 'px' })
    }

    return (
      <Form>
        <Input
          placeholder="Send a message..."
          value={this.state.value}
          innerRef={el => {
            this.autogrow = el
          }}
          style={style}
          isPinned={isPinned}
          onChange={e => this.handleChange(e)}
          onKeyUp={e => this.handleKeyUp(e)}
        />
      </Form>
    )
  }
}
