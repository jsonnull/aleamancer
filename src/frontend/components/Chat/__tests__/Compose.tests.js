// @flow
import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import sinon from 'sinon'
import Compose from '../Compose.js'

describe('Chat Compose component', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(<Compose onSend={() => {}} isPinned={false} />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should handle ENTER event', () => {
    const onSend = sinon.spy()
    const wrapper = mount(<Compose onSend={onSend} isPinned={false} />)
    wrapper.find('textarea').simulate('keyUp', { key: 'Enter' })
    expect(onSend.called).toBe(true)
  })
})
