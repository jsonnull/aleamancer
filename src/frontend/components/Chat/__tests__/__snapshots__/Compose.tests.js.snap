// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat Compose component renders correctly 1`] = `
.c0 {
  border: 0;
  height: 50px;
  line-height: 3rem;
  padding: 1rem;
  display: block;
  width: 100%;
  resize: none;
  border-radius: 5px;
  font-size: 1.5rem;
  font-family: -apple-system,BlinkMacSystemFont,  "Segoe UI","Roboto","Oxygen","Ubuntu","Cantarell","Fira Sans","Droid Sans",  "Helvetica Neue",sans-serif;
}

<form
  className=""
>
  <textarea
    className="c0"
    onChange={[Function]}
    onKeyUp={[Function]}
    placeholder="Send a message..."
    style={Object {}}
    type="text"
    value=""
  />
</form>
`;
